### PR TITLE
Add ability to decode Message Annotations

### PR DESCRIPTION
**Pull request by FedericoDiTosto**

### CODE REVIEW

**Review by albertobarrila — APPROVED** [review]
I've just cleaned the tests that refers always to the message list while not using it

File: src/amqp10/encoder.ts [hunk 1]
```diff
@@ -34,14 +35,21 @@ const PropertySizeDescription =
 
 type MessageApplicationPropertiesList = [string, string | number][]
 
-export function amqpEncode(writer: DataWriter, { content, messageProperties, applicationProperties }: Message): void {
+type MessageAnnotationsList = [string, string | number][]
```
Comment (gpad) [inline src/amqp10/encoder.ts:38]: I don't have seen this list of list before why don't create something more usefull like:
```
{key: string, value: string| number}[]
```
if you agree I suggest to refactor also the `MessageApplicationPropertiesList`

File: src/amqp10/messageAnnotations.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  public static parse(dataReader: DataReader, elementsLength: number): MessageAnnotations {
+    const numEntries = elementsLength / 2
+
+    return range(numEntries).reduce((acc: MessageAnnotations, _) => {
+      const propertyKey = readUTF8String(dataReader)
+      const nextByteType = dataReader.readUInt8()
+      dataReader.rewind(1)
+      const propertyValue = decodeFormatCode(dataReader, nextByteType, true)
+      if (!propertyValue) throw new Error(`invalid nextByteType %#02x: ${nextByteType}`)
+      acc[propertyKey] = propertyValue as string | number
```
Comment (gpad) [inline src/amqp10/messageAnnotations.ts:16]: Are you sure about this `as` ?

File: test/e2e/declare_consumer.test.ts [hunk 3]
```diff
@@ -82,13 +88,11 @@ describe("declare consumer", () => {
...
   it("declaring a consumer on an existing stream - the consumer should read message properties", async () => {
     const messageProperties: MessageProperties[] = []
-    const messages: string[] = []
     const properties = createProperties()
     await publisher.send(Buffer.from("hello"), { messageProperties: properties })
 
     await connection.declareConsumer({ stream: streamName, offset: Offset.first() }, (message: Message) => {
-      messageProperties.push(message.messageProperties || {})
-      messages.push("JSON.stringify(message.properties?.correlationId) ||")
+      if (message.messageProperties) messageProperties.push(message.messageProperties)
```
Comment (gpad) [inline test/e2e/declare_consumer.test.ts:95]: Why have you changed this test?

File: test/e2e/declare_consumer.test.ts [hunk 4]
```diff
@@ -98,36 +102,48 @@ describe("declare consumer", () => {
...
     })
   }).timeout(10000)
+
+  it("declaring a consumer on an existing stream - the consumer should read message annotations", async () => {
+    const messageAnnotations: MessageAnnotations[] = []
+    const annotations = createAnnotations()
+    await publisher.send(Buffer.from("hello"), { messageAnnotations: annotations })
+
+    await connection.declareConsumer({ stream: streamName, offset: Offset.first() }, (message: Message) => {
+      if (message.messageAnnotations) messageAnnotations.push(message.messageAnnotations)
```
Comment (gpad) [inline test/e2e/declare_consumer.test.ts:123]: I don't think this iff is necessary, if you get an error from typescript you can solve it like this
```suggestion
messageAnnotations.push(message.messageAnnotations || {})
```

File: test/e2e/declare_consumer.test.ts [hunk 5]
```diff
@@ -98,36 +102,48 @@ describe("declare consumer", () => {
...
   it("declaring a consumer on an existing stream - the consumer should read application properties", async () => {
     const messageApplicationProperties: MessageApplicationProperties[] = []
-    const messages: string[] = []
     const applicationProperties = createApplicationProperties()
     await publisher.send(Buffer.from("hello"), { applicationProperties })
 
     await connection.declareConsumer({ stream: streamName, offset: Offset.first() }, (message: Message) => {
-      messageApplicationProperties.push(message.applicationProperties || {})
-      messages.push("JSON.stringify(message.properties?.correlationId) ||")
+      if (message.applicationProperties) messageApplicationProperties.push(message.applicationProperties)
```
Comment (gpad) [inline test/e2e/declare_consumer.test.ts:109]:
```suggestion
      messageApplicationProperties.push(message.applicationProperties || {})
```

File: test/e2e/declare_consumer.test.ts (hunk 3 above)
Comment (gpad) [reply test/e2e/declare_consumer.test.ts:95]:
```suggestion
messageProperties.push(message.messageProperties|| {})
```

File: test/e2e/declare_consumer.test.ts [hunk 6]
```diff
@@ -98,36 +102,48 @@ describe("declare consumer", () => {
...
+    const annotations = createAnnotations()
+    await publisher.send(Buffer.from("hello"), { messageAnnotations: annotations })
+
+    await connection.declareConsumer({ stream: streamName, offset: Offset.first() }, (message: Message) => {
+      if (message.messageAnnotations) messageAnnotations.push(message.messageAnnotations)
+    })
+
+    await eventually(async () => {
+      expect(messageAnnotations).eql([annotations])
+    })
```
Comment (gpad) [inline test/e2e/declare_consumer.test.ts:128]:
```suggestion
    await eventually(async () => expect(messageAnnotations).eql([annotations]))
```

File: test/e2e/declare_consumer.test.ts
Comment (gpad) [inline test/e2e/declare_consumer.test.ts:1]: Can you double-check if the message annotation is ignored by the classic driver? Just to be sure, we are make a more useful test.

File: src/amqp10/messageAnnotations.ts (hunk 2 above)
Comment (l4mby) [reply src/amqp10/messageAnnotations.ts:16]: I don't know how to do it otherwise, decodeFormatCode has multiple return types

**Review by gpad — APPROVED** [review]

**Pull request merged**